packages/flutter/lib/src/painting/images.dart

```diff
@@ -157,6 +166,106 @@ class DecorationImage {
   }
 }
 
+/// The painter for a [DecorationImage].
+///
+/// To obtain a painter, call [DecorationImage.createPainter].
+///
+/// To paint, call [paint]. The `onChanged` callback passed to
+/// [DecorationImage.createPainter] will be called if the image needs to paint
+/// again (e.g. because it is animated or because it had not yet loaded the
+/// first time).
+///
+/// This object should be disposed using the [dispose] method when it is no
+/// longer needed.
+class DecorationImagePainter {
+  DecorationImagePainter._(this._details, this._onChanged);
+
+  final DecorationImage _details;
+  final VoidCallback _onChanged;
+
+  ImageStream _imageStream;
+  ImageInfo _image;
+
+  void paint(Canvas canvas, Rect rect, Path clipPath, ImageConfiguration configuration) {
```

Comment:
It might be useful to assert that `dispose` hasn't been called first thing in this method.

Reply:
Short of adding a field just for that purpose, there's not an easy way I can think to do that.

Reply:
A debug field SGTM, unless you're worried about performance.

Reply:
I'm inclined to not worry about this until we find that this is actually a problem people are running into. We do occasionally worry about this, but only for particularly high-profile cases. In this case, this API is mostly an internal implementation detail. Few people are ever going to use it directly.